collectors/proc.plugin/sys_class_power_supply.c

```diff
@@ -245,67 +245,75 @@ int do_sys_class_power_supply(int update_every, usec_t dt) {
+                        power_supply_free(ps);
+                        ps = NULL;
+                    }
+                    else {
+                        buffer[r] = '\0';
+                        ps->capacity->value = str2ull(buffer);
 
-                if(unlikely(!keep_fds_open)) {
-                    close(ps->capacity->fd);
-                    ps->capacity->fd = -1;
-                }
-                else if(unlikely(lseek(ps->capacity->fd, 0, SEEK_SET) == -1)) {
-                    error("Cannot seek in file '%s'", ps->capacity->filename);
-                    close(ps->capacity->fd);
-                    ps->capacity->fd = -1;
+                        if(unlikely(!keep_fds_open)) {
+                            close(ps->capacity->fd);
+                            ps->capacity->fd = -1;
+                        }
+                        else if(unlikely(lseek(ps->capacity->fd, 0, SEEK_SET) == -1)) {
+                            error("Cannot seek in file '%s'", ps->capacity->filename);
+                            close(ps->capacity->fd);
+                            ps->capacity->fd = -1;
+                        }
+                    }
                 }
             }
 
             // read property files
             int read_error = 0;
             struct ps_property *pr;
-            for(pr = ps->property_root; pr && !read_error; pr = pr->next) {
-                struct ps_property_dim *pd;
-                for(pd = pr->property_dim_root; pd; pd = pd->next) {
-                    if(likely(!pd->always_zero)) {
-                        char buffer[30 + 1];
-
-                        if(unlikely(pd->fd == -1)) {
-                            pd->fd = open(pd->filename, O_RDONLY, 0666);
+            if (ps)
+            {
+                for(pr = ps->property_root; pr && !read_error; pr = pr->next) {
+                    struct ps_property_dim *pd;
+                    for(pd = pr->property_dim_root; pd; pd = pd->next) {
+                        if(likely(!pd->always_zero)) {
+                            char buffer[30 + 1];
+
                             if(unlikely(pd->fd == -1)) {
-                                error("Cannot open file '%s'", pd->filename);
+                                pd->fd = open(pd->filename, O_RDONLY, 0666);
+                                if(unlikely(pd->fd == -1)) {
+                                    error("Cannot open file '%s'", pd->filename);
+                                    read_error = 1;
+                                    power_supply_free(ps);
+                                    break;
+                                }
+                            }
+
+                            ssize_t r = read(pd->fd, buffer, 30);
+                            if(unlikely(r < 1)) {
+                                error("Cannot read file '%s'", pd->filename);
                                 read_error = 1;
                                 power_supply_free(ps);
```

Comment:
This does not break cleanly out of the 2 nested loops and it can be called multiple times in theory. That would be `free()` after `free()` which sounds bad.

Reply:
It was fixed now.

Reply:
But still the code can go to `pr = pr->next` from:
```
for(pr = ps->property_root; pr && !read_error; pr = pr->next) {
```
and call `power_supply_free(ps);` another time?

Reply:
The `read_error` variable avoid it continues going in the next element of the for loop, as you can see it is set to `1`, every time we clean the structure.

Reply:
Ah, yes, you're right!

Reply:
Then I guess `ps = NULL;` is not really necessary because the next use is only `struct power_supply *ps = power_supply_root;` ?

Reply:
I agree 100%, I removed the unnecessary set of a variable. Thank you!